### PR TITLE
libsubprocess: support user friendly error string 

### PR DESCRIPTION
**Pull request by chu11**
Problem: When a subprocess fails the "failed_errno" value is set and can be returned in `flux_subprocess_fail_errno()`.  It would
be convenient if there was an error message to describe the error in more detail.

Solution: Internally support a new "failed_error" field which can store a text message describing the error in more detail.  Generally speaking, always fill this with a text message when "failed_errno" is also set.

Set user friendly error string in many cases throughout libsubprocess.

Support a new `flux_subprocess_fail_error()` to retrieve this error message.

Add corner case unit tests.

Fixes https://github.com/flux-framework/flux-core/issues/5233

----

Extra notes:

I think the biggest review question in this PR will be my choice of "error message" throughout.  And if I should add it in more places.  We capture the big ones of EACCESS / EPERM on a remotely executed process. e.g.

```
flux-exec: Error: rank 0: ./nosuchprocess: error launching process: No such file or directory
```

### CODE REVIEW

**Comment by chu11** [comment]
oops, spelling error, quick fix, squash, re-push

File: src/common/libsubprocess/subprocess.c [hunk 1]
```diff
@@ -1121,6 +1121,21 @@ int flux_subprocess_fail_errno (flux_subprocess_t *p)
...
+        return NULL;
+    }
+    if (p->state != FLUX_SUBPROCESS_FAILED) {
+        errno = EINVAL;
+        return NULL;
+    }
+    if (p->failed_error.text[0] == '\0')
+        return strerror (p->failed_errno);
+    return p->failed_error.text;
+}
```
Comment (garlick) [inline src/common/libsubprocess/subprocess.c:1137]: I'm not sure I know all the places where this will be used but sometimes it's inconvenient to return NULL from a function like this.  Would it be better to to just return textual errors like
- `internal error: subprocess is NULL`
- `internal error: subprocess is not in FAILED state`

then the user can just use the result in a printf argument without checking?

**Review by garlick — APPROVED** [review]
:tada: 

Nice work!  This will be helpful.  I had a few comments inline, questions really.  But this generally looks great.  Thanks!

File: src/common/libsubprocess/client.c [hunk 2]
```diff
@@ -130,7 +130,7 @@ flux_future_t *subprocess_rexec (flux_t *h,
     return NULL;
 }
 
-int subprocess_rexec_get (flux_future_t *f)
+int subprocess_rexec_get (flux_future_t *f, flux_error_t *error)
```
[the remark this answers is not in the document]
Comment (chu11) [reply src/common/libsubprocess/client.c:133]: sounds good.  I had it originally but thought "ehhh there could be more detail", guess I went too far.

File: src/common/libsubprocess/client.c [hunk 3]
```diff
@@ -157,7 +157,7 @@ int subprocess_rexec_get (flux_future_t *f)
     }
     else if (!streq (ctx->response.type, "started")
         && !streq (ctx->response.type, "stopped")
-        && !streq (ctx->response.type, "finished")) {
+             && !streq (ctx->response.type, "finished")) {
```
Comment (garlick) [inline src/common/libsubprocess/client.c:160]: whoops?

**Review by garlick — APPROVED** [review]
A typo crept in?  o/w looks good!

File: src/common/libsubprocess/client.c (hunk 3 above)
Comment (chu11) [reply src/common/libsubprocess/client.c]: doh! good catch

**Comment by chu11** [comment]
re-pushed with fixes per comments above, also ...

- I added an error message for "internal fatal" errors, perhaps is unnecessary but just in case.

```C
@@ -136,6 +137,7 @@ static void proc_internal_fatal (flux_subprocess_t *p)
      */
     p->state = FLUX_SUBPROCESS_FAILED;
     p->failed_errno = errno;
+    errprintf (&p->failed_error, "internal fatal error: %s", strerror (errno));
     state_change_start (p);
```

- I think an error message of "NULL" is different than "empty string", so did this tweak to what was there before

```C
     else if (state == FLUX_SUBPROCESS_FAILED) {
+        const char *errmsg = NULL;
+        if (p->failed_error.text[0] != '\0')
+            errmsg = p->failed_error.text;
         rc = flux_respond_error (s->h,
                                  request,
                                  p->failed_errno,
-                                 p->failed_error.text);
+                                 errmsg);
```

alternately I could call `flux_subprocess_fail_errno(p)` and `flux_subprocess_fail_error(p)` so there would always be a message to
send.

- I noticed English errors in the tests, where I said "inputs" where it should be "input".  Dumb new commit added.
